gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/monitor/ClouderaManagerClusterConfigurationMonitor.java

```diff
@@ -0,0 +1,876 @@
+    ServicesResourceApi api = new ServicesResourceApi(apiClient);
+    try {
+      ApiServiceConfig svcConfig = api.readServiceConfig(clusterName, service, "full");
+
+      Map<ApiRole, ApiConfigList> roleConfigs = new HashMap<>();
+      RolesResourceApi rolesApi = (new RolesResourceApi(apiClient));
+      ApiRoleList roles = rolesApi.readRoles(clusterName, service, "", "full");
+      for (ApiRole role : roles.getItems()) {
+        ApiConfigList config = rolesApi.readRoleConfig(clusterName, role.getName(), service, "full");
+        roleConfigs.put(role, config);
+      }
+      currentConfig = new ServiceConfigurationModel(svcConfig, roleConfigs);
+    } catch (ApiException e) {
+      log.clouderaManagerConfigurationAPIError(e);
+    }
+    return currentConfig;
+  }
+
+  /**
+   * Get restart events for the specified ClouderaManager cluster.
+   *
+   * @param address     The address of the ClouderaManager instance.
+   * @param clusterName The name of the cluster.
+   * @return A List of RestartEvent objects for service restart events since the last time they were queried.
+   */
+  private List<RestartEvent> getRestartEvents(final String address, final String clusterName) {
+    List<RestartEvent> restartEvents = new ArrayList<>();
+
+    // Get the last event query timestamp
+    String lastTimestamp = getEventQueryTimestamp(address, clusterName);
+
+    if (lastTimestamp == null) {
+      Date ts = new Date();
+      ts.setTime(System.currentTimeMillis() - eventQueryDefaultTimestampOffset);
+      lastTimestamp = eventQueryTimestampFormat.get().format(ts);
+    }
+
+    log.queryingRestartEventsFromCluster(clusterName, address, lastTimestamp);
+
+    // Record the new event query timestamp for this address/cluster
+    setEventQueryTimestamp(address, clusterName, new Date());
+
+    // Query the event log from CM for service/cluster restart events
+    List<ApiEvent> events = queryRestartEvents(getApiClient(getDiscoveryConfig(address, clusterName)),
+        clusterName,
+        lastTimestamp);
+    for (ApiEvent event : events) {
+      restartEvents.add(new RestartEvent(event));
+    }
+
+    return restartEvents;
+  }
+
+  private void persist(final Properties props, final File dest) {
+    try (OutputStream out = Files.newOutputStream(dest.toPath())) {
+      props.store(out, PERSISTED_FILE_COMMENT);
+      out.flush();
+    } catch (Exception e) {
+      log.failedToPersistClusterMonitorData(getType(), dest.getAbsolutePath(), e);
+    }
+  }
+
+  private File getPersistenceDir() {
+    File persistenceDir = null;
+
+    File dataDir = new File(gatewayConfig.getGatewayDataDir());
+    if (dataDir.exists()) {
+      File clustersDir = new File(dataDir, CLUSTERS_DATA_DIR_NAME);
+      if (!clustersDir.exists()) {
+        if (!clustersDir.mkdirs()) {
+          log.failedToCreatePersistenceDirectory(clustersDir.getAbsolutePath());
+        }
+      }
+      persistenceDir = clustersDir;
+    }
+
+    return persistenceDir;
+  }
```

Comment:
Nit: It might be cleaner to use `Path` and Java NIO here. This would cascade to the rest of the `File` usages.